File: cpp/src/phase1.cpp
#include "simplex/phase1.hpp"
#include "simplex/preprocess.hpp"
#include <Eigen/LU>
#include <stdexcept>

// Forward declaration for the helper
static void fix_artificials_in_basis(Eigen::MatrixXd& Aw,
                                      std::vector<int>& basic,
                                      std::vector<int>& nonbasic,
                                      int art_start);

std::tuple<BasisState, SolveStatus>
phase1(const LinearProgram& lp_in, const PivotRule& rule,
       const SimplexLogger& logger, int maxit, double tol)
{
    LinearProgram lp = make_b_nonnegative(lp_in);
    int m = static_cast<int>(lp.A.rows());
    int n = static_cast<int>(lp.A.cols());

    // Aw = [A | I_m], artificials start at column index n (0-based)
    Eigen::MatrixXd Aw(m, n + m);
    Aw.leftCols(n)  = lp.A;
    Aw.rightCols(m) = Eigen::MatrixXd::Identity(m, m);

    Eigen::VectorXd cw(n + m);
    cw.head(n).setZero();
    cw.tail(m).setConstant(-1.0);

    int art_start = n; // 0-based

    std::vector<int> basic(m), nonbasic(n);
    for (int i = 0; i < m; ++i) basic[i]    = art_start + i;
    for (int i = 0; i < n; ++i) nonbasic[i] = i;

    logger.log_phase(1);

    for (int it = 1; it <= maxit; ++it) {
        // Build B and N from current basis
        Eigen::MatrixXd B(m, m), N(m, n);
        for (int j = 0; j < m; ++j) B.col(j) = Aw.col(basic[j]);
        for (int j = 0; j < n; ++j) N.col(j) = Aw.col(nonbasic[j]);

        Eigen::PartialPivLU<Eigen::MatrixXd> BF(B);
        Eigen::VectorXd xb    = BF.solve(lp.b);
        Eigen::MatrixXd BinvN = BF.solve(N);

        Eigen::VectorXd cb(m);
        for (int i = 0; i < m; ++i) cb[i] = cw[basic[i]];

        Eigen::VectorXd cr_nb(n);
        for (int j = 0; j < n; ++j) cr_nb[j] = cw[nonbasic[j]];
        cr_nb -= BinvN.transpose() * cb;

        double z = cb.dot(xb);

        logger.log_iteration(it, BasisState{basic, nonbasic}, xb, z);

        auto opt_j = rule.entering_index(cr_nb, tol);

        if (!opt_j.has_value()) {
            if (z < -tol) {
                return { BasisState{{}, {}}, SolveStatus::INFEASIBLE };
            }
            fix_artificials_in_basis(Aw, basic, nonbasic, art_start);

            std::vector<int> orig_basic, orig_nonbasic;
            for (int idx : basic)    if (idx < n) orig_basic.push_back(idx);
            for (int idx : nonbasic) if (idx < n) orig_nonbasic.push_back(idx);
            return { BasisState{orig_basic, orig_nonbasic}, SolveStatus::OPTIMAL };
        }

        int j = *opt_j;
        Eigen::VectorXd d = BinvN.col(j);
        auto opt_i = leaving_index(xb, d, BinvN, tol);

        if (!opt_i.has_value())
            throw std::runtime_error("Phase 1 is unexpectedly unbounded");

        int i = *opt_i;
        std::swap(basic[i], nonbasic[j]);
    }

    throw std::runtime_error("Phase 1 did not converge in " +
                              std::to_string(maxit) + " iterations");
}

static void fix_artificials_in_basis(Eigen::MatrixXd& Aw,
                                      std::vector<int>& basic,
                                      std::vector<int>& nonbasic,
                                      int art_start)
{
    int m = static_cast<int>(basic.size());
    bool changed = true;
    while (changed) {
        changed = false;
        for (int pos = 0; pos < m; ++pos) {
            if (basic[pos] < art_start) continue;

            Eigen::MatrixXd B(Aw.rows(), m);
            for (int j = 0; j < m; ++j) B.col(j) = Aw.col(basic[j]);
            Eigen::PartialPivLU<Eigen::MatrixXd> BF(B);
            Eigen::MatrixXd BinvAw = BF.solve(Aw);
            Eigen::VectorXd row = BinvAw.row(pos).transpose();

            int swap_j = -1;
            for (int j = 0; j < static_cast<int>(nonbasic.size()); ++j) {
                int ncol = nonbasic[j];
                if (ncol < art_start && std::abs(row[ncol]) > 1e-10) {
                    swap_j = j;
                    break;
                }
            }

            if (swap_j < 0) {
                Aw.row(pos).setZero();
            } else {
                std::swap(basic[pos], nonbasic[swap_j]);
                changed = true;
                break;
            }
        }
    }
}

File: typescript/src/phase1.ts
import type { LinearProgram, BasisState } from './types.js';
import type { SolveStatus } from './types.js';
import type { PivotRule } from './pivot.js';
import { enteringIndex, leavingIndex } from './pivot.js';
import { luFactor, luSolve, luSolveMatrix, dot } from './linalg.js';
import { getColumns } from './linalg.js';
import { makeBNonnegative } from './preprocess.js';
import type { SimplexLogger } from './logger.js';

export function phase1(
  lpIn: LinearProgram,
  rule: PivotRule,
  logger: SimplexLogger,
  maxit: number = 1000,
  tol: number = 1e-8,
): [BasisState, SolveStatus] {
  const lp = makeBNonnegative(lpIn);
  const m = lp.A.length;
  const n = lp.A[0]?.length ?? 0;

  // Augmented matrix Aw = [A | I_m], 0-based column indices
  // Artificial variables occupy columns n..n+m-1
  const artStart = n; // 0-based start of artificial columns
  const Aw: number[][] = lp.A.map((row, i) => {
    const ext = new Array(m).fill(0);
    ext[i] = 1;
    return [...row, ...ext];
  });

  // Objective for phase 1: minimize sum of artificials -> maximize cw
  // cw[0..n-1] = 0, cw[n..n+m-1] = -1
  const cw: number[] = [...new Array(n).fill(0), ...new Array(m).fill(-1)];

  // Initial basis: artificials are basic, original vars are nonbasic
  const basic: number[] = Array.from({ length: m }, (_, i) => artStart + i);
  const nonbasic: number[] = Array.from({ length: n }, (_, j) => j);

  logger.logPhase(1);

  for (let it = 1; it <= maxit; it++) {
    const B = getColumns(Aw, basic);
    const N = getColumns(Aw, nonbasic);
    const BF = luFactor(B);
    const xb = luSolve(BF, lp.b);
    const BinvN = luSolveMatrix(BF, N);

    const cb = basic.map(j => cw[j]);
    const cr = nonbasic.map((j, col) => {
      let s = cw[j];
      for (let row = 0; row < cb.length; row++) s -= BinvN[row][col] * cb[row];
      return s;
    });
    const z = dot(cb, xb);

    logger.logIteration(it, { basic: basic.slice(), nonbasic: nonbasic.slice() }, xb, z);

    const j = enteringIndex(rule, cr, tol);
    if (j === null) {
      // Phase 1 terminated; check feasibility
      if (z < -tol) return [{ basic: [], nonbasic: [] }, 'INFEASIBLE'];

      fixArtificialsInBasis(Aw, basic, nonbasic, artStart);

      const origBasic = basic.filter(i => i < artStart);
      const origNonbasic = nonbasic.filter(i => i < artStart);
      return [{ basic: origBasic, nonbasic: origNonbasic }, 'OPTIMAL'];
    }

    const d = BinvN.map(row => row[j]);
    const i = leavingIndex(xb, d, BinvN, tol);

    // Phase 1 auxiliary problem is always bounded
    if (i === null) throw new Error('Phase 1 is unexpectedly unbounded');

    const tmp = basic[i];
    basic[i] = nonbasic[j];
    nonbasic[j] = tmp;
  }

  throw new Error(`Phase 1 did not converge in ${maxit} iterations`);
}

function fixArtificialsInBasis(
  Aw: number[][],
  basic: number[],
  nonbasic: number[],
  artStart: number,
): void {
  let changed = true;
  while (changed) {
    changed = false;
    for (let pos = 0; pos < basic.length; pos++) {
      if (basic[pos] < artStart) continue;

      const B = getColumns(Aw, basic);
      const BF = luFactor(B);
      // BinvAw = B^{-1} * Aw  (solve for each column of Aw)
      const BinvAw = luSolveMatrix(BF, Aw);
      const row = BinvAw[pos];

      let swapJ: number | null = null;
      for (let j = 0; j < nonbasic.length; j++) {
        const ncol = nonbasic[j];
        if (ncol < artStart && Math.abs(row[ncol]) > 1e-10) {
          swapJ = j;
          break;
        }
      }

      if (swapJ === null) {
        // Row is linearly dependent; zero it out to remove degenerate artificial
        for (let col = 0; col < Aw[pos].length; col++) Aw[pos][col] = 0;
      } else {
        const tmp = basic[pos];
        basic[pos] = nonbasic[swapJ];
        nonbasic[swapJ] = tmp;
        changed = true;
        break;
      }
    }
  }
}
